File: balance binary tree.py
# time: O(n) and O(1) if we not considering stack space
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        flag = True
        def solve(root):
            nonlocal flag
            if root is None:
                return 0
            left = solve(root.left)
            right = solve(root.right)
            if abs(left - right) > 1:
                flag = False
            return 1 + max(left, right)
        solve(root)
        return flag

File: balance binary tree.cpp
class Solution {
public:
    int solve(TreeNode * root, bool &flag){
        if (root == nullptr){
            return 0;
        }
        int left = solve(root->left, flag);
        int right = solve(root->right, flag);
        if (abs(left - right) > 1) flag = false;
        return 1 + max(left, right);
    }
    bool isBalanced(TreeNode* root) {
        bool flag = true;
        solve(root, flag);
        return flag;
    }
};
